docs/_posts/2015-02-24-streamlining-react-elements.md

```diff
@@ -0,0 +1,230 @@
+```
+
+### Problem: Relies on Enumeration Order
+
+The problem with this pattern is that it relies on enumeration order of objects. This is technically unspecified, even though implementations now agree to use insertion order. Except for the special case when numeric keys are used.
+
+### Problem: Using Objects as Maps is Bad
+
+It is generally accepted that using objects as maps screw up type systems, VM optimizations, compilers etc. It is my better to use a dedicated data structure like ES6 Maps.
+
+More importantly, this can have important security implications. For example this has a potential security problem:
+
+```js
+var children = {};
+items.forEach(item => x[item.title] = <span />);
+return <div>{children}</div>;
+```
+
+Imagine if `item.title === '__proto__'` for example.
+
+### Problem: Can’t be Differentiated from Arbitrary Objects
+
+Since these objects can have any keys with almost any value, we can’t differentiate them from a mistake. If you put some random object, we will try our best to traverse it and render it, instead of failing with a helpful warning. In fact, this is one of the few places where you can accidentally get an infinite loop in React.
+
+To differentiate ReactElements from one of these objects, we have to tag them with `_isReactElement` which we wouldn’t need to without this.
+
+### Solution: Just use an Array and key={…}
+
+Most of the time you can just use an array with keyed ReactElements.
+
+```js
+var children = items.map(item => <span key={item.title} />);
+<div>{children}</div>
+```
+
+### Solution: React.addons.createFragment
+
+However, this is not always possible if you’re trying to add a prefix key to an unknown set (e.g. this.props.children). It is also not always the easiest upgrade path. Therefore, we are adding a helper to `React.addons` called `createFragment()`. This accepts a keyed object and returns an opaque type.
+
+```js
+<div>{React.addons.createFragment({ a: <div />, b: this.props.children })}</div>
+```
+
+The exact signature of this kind of fragment will be determined later. It will likely be some kind of immutable sequence.
+
+Note: This will still not be valid as the direct return value of `render()`. Unfortunately, they still need to be wrapped in a `<div />` or some other element.
+
+## Compiler Optimizations: Unlocked!
+
+These changes also unlock several possible compiler optimizations for static content in React 0.14. These optimizations were previously only available to template-based frameworks. They will now also be possible for React code! Both for JSX and `React.createElement/Factory`*!
+
+- [Reuse Constant Value Types](https://github.com/facebook/react/issues/3226)
+
+- [Tagging ReactElements](https://github.com/facebook/react/issues/3227)
+
+- [Inline ReactElements](https://github.com/facebook/react/issues/3228)
+
+*) If you use the recommended pattern of explicit React.createFactory calls on the consumer side - since they are easily statically analyzed.
```

Comment:
Might not actually be worth quantifying that. You could instead say in the sentence this asterisk came from "both for JSX and idiomatic uses of createElement/Factory"